alternator: close output_stream when exception is thrown during response streaming

nuivall opened the pull request:
When exception occurs and we omit closing output_stream then the whole process is brought down by an assertion in ~output_stream.

Fixes https://github.com/scylladb/scylladb/issues/14453 
Relates https://github.com/scylladb/scylladb/issues/14403

nuivall commented:
It's quite hard to write automated test for the issue but I felt it's important because of the severity of the problem, so I came up with something quite hacky - but at least it's works reliably for me.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2161/

nyh commented on alternator/executor.cc:
```diff
@@ -109,16 +109,20 @@ json::json_return_type make_streamed(rjson::value&& value) {
+            ex = std::current_exception();
+            elogger.error("Exception during streaming HTTP response: {}", ex);
+        }
+        co_await los.close();
```
Good catch! I have to say I find it very surprising/upsetting that after writing to a file has already **failed**, it is still mandatory to close it otherwise you get an assertion from Seastar. It's like a bug magnet. But even if that Seastar issue was somehow solved (which it isn't), your patch is still good and important because in theory rjson::print() could have failed on something else - not just a file error.

nyh commented on test/alternator/conftest.py:
```diff
@@ -60,14 +60,7 @@ def pytest_collection_modifyitems(config, items):
-# We use scope="session" so that all tests will reuse the same client object.
-@pytest.fixture(scope="session")
-def dynamodb(request):
+def _dynamodb(request):
```
An idea: Instead of the ugly name "_dynamodb" which the user is not supposed to use (you created dynamodb_function_scoped instead), you could have a useful context manager, say "dynamodb_session", which the user can use directly and then you don't need dynamodb_function_scoped. In cql-pytest I did something like this - there is (in util.py) a cql_session() context manager, which tests can use directly (or through the new_session() wrapper) and the "cql" fixture also uses.

nuivall replied:
looks like closing is not required in boto3, it closes connections when object is destructed, couldn't find precise docs about it but at least people on the internet say so...

nyh commented on test/alternator/conftest.py:
```diff
@@ -89,6 +82,21 @@ def dynamodb(request):
+# "dynamodb" fixture for tests which want their own client instance.
+@pytest.fixture(scope="function")
+def dynamodb_function_scoped(request):
+    return _dynamodb(request)
```
I'm worried that the fact that these fixtures all "return" the resource instead of yield+cleanup, and that the new _dynamodb is a function and not a contextmanager, was a mistake on my part when I created the dynamodb fixture. Shouldn't we need to close() the resource? I'm worried that it didn't matter when we had just one "dynamodb" fixture for the entire run, but will start mattering when we have all those temporary sessions.

nuivall replied:
ok, switched to non fixture func

nyh commented on test/alternator/test_batch.py:
```diff
@@ -375,6 +378,43 @@ def test_batch_write_item_large(test_table_sn):
+                nonlocal triggered
+                triggered = True
+            return ret
+        urllib3.HTTPResponse.read_chunked = broken_read_fun
```
You said that you used a separate dynamodb "resource" because of this monkey-patching. But I don't understand how the separate resource helps... You're monkey-patching a global urllib3 function, which all boto3 code supposedly uses, in all sessions. What am I missing?

nuivall replied:
Two things are connected to this:
- I hacked this breaking patch with `st = traceback.extract_stack()` to inject error only if function name is found in the trace.
- I noticed dynamodb instance can't recover after error is injected, but other instance/session I think won't share connection pools.

nyh commented on test/alternator/test_batch.py:
```diff
@@ -375,6 +378,43 @@ def test_batch_write_item_large(test_table_sn):
+        urllib3.HTTPResponse.read_chunked = broken_read_fun
+
+        try:
+            # Our monkey patched read_chunked function will make client unusable
```
You said that you used a separate dynamodb "resource" because of this monkey-patching. But I don't understand how the separate resource helps... You're monkey-patching a global urllib3 function, which all boto3 code supposedly uses, in all sessions.
Maybe you meant that you need the new session because of the `close()` call **inside** the monkey-patched function, not because of the monkey-patching?

nuivall replied:
yes, as wrote "monkey patched read_chunked function will make client unusable", the function `read_chunked` breaks the session (due to close but that's detail)

nuivall commented:
v2:
- tests: switched to non fixture func for new session

nyh commented:
@nuivall I tried running your test **without** the code patch, and it doesn't reproduce the bug (the test passes, and Syclla doesn't crash). Increasing the loop count from 20 to 1000 didn't help - all 1000 iterations passed. Increasing the size of long_content 3-fold also didn't help reproduce a bug, and increasing it 10-fold caused a write failure, but not the crash I expected. 

I was running this test with test/alternator/run. Perhaps you were running it differently, maybe over a slower network? Were you able to reproduce a crash?

nuivall commented on test/alternator/conftest.py:
```diff
@@ -89,6 +90,16 @@ def dynamodb(request):
+        conf = conf.merge(botocore.client.Config(retries={"max_attempts": 0}, read_timeout=300))
+    return ses.resource('dynamodb', endpoint_url=dynamodb.meta.client._endpoint.host, verify=host.scheme != 'http',
+        region_name='us-east-1', aws_access_key_id='alternator', aws_secret_access_key='secret_pass',
+        config=conf)
```
This piece of code is inheriting configs from dynamodb, one thing I forgot is to return early if it's not localhost

nyh replied:
I don't think this code is inheriting anything from the "dynamodb" fixture, it creates a new config object.

nyh commented:
In any case, we **could** commit this fix without a reliable test, because the code makes sense and you were able to test it yourself. I just don't want to include in the test suite a test that doesn't actually work. Maybe we can insert it into the code with "@skip" saying it isn't reliable - and if we want, we can work on improving it later... Or should we just not include this test?

nuivall commented:
I think it broke after refactoring, new_session  should be outside of the loop, let me check

nuivall commented:
v3:
- fixed test, broken after refactor
- fixed aws handling in new_dynamodb_session (it inherits settings from dynamodb, but it can't easily know if this is aws or not so needs to use request fixture too)

nyh commented on test/alternator/test_batch.py:
```diff
@@ -375,6 +379,43 @@ def test_batch_write_item_large(test_table_sn):
+    for _ in range(1, 20):
+        p = random_string()
+        long_content = random_string(100)*500
+        write_reply = test_table_sn.meta.client.batch_write_item(RequestItems = {
```
By the way, is writing the same item again and again 20 times part of this test, or it's enough to just read it 20 times?

nuivall replied:
good point, only read is required, will limit the loop

nyh commented:
Unfortunately, even with the new version, I can't seem to get the test to ever fail without your code patch...

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2203/

nuivall commented:
> Unfortunately, even with the new version, I can't seem to get the test to ever fail without your code patch...

That's odd, it reproduces 100% on my machine. I have scylla master ff386e7a445d17ed2199a26e4ac75905b071a91b and boto3:
```
❯ pip show boto3
Name: boto3
Version: 1.20.26
```

nuivall commented:
v4: 
- test: minimize scope of retry loop

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2230/

nyh commented:
As usual some sort of weird CI failure I don't understand (no specific test which failed), so I'll restart it.

nyh commented on test/alternator/test_batch.py:
```diff
@@ -375,6 +379,43 @@ def test_batch_write_item_large(test_table_sn):
+
+    try:
+        # This disruption doesn't always work so we repeat it.
+        for _ in range(1, 20):
```
I don't understand what this loop intends to achieve over running once. It checks that on **every iteration**, the disruption happened - that the query must raise HTTPClientError every time, and triggered must be true after each iteration (by the way, after the first time, it just remains true! maybe you meant to set it to false every time?).

What I'm guessing you really wanted to say in the comment above the loop was that before your patch, Scylla only crashed randomly so you needed to run it 20 times to crash more reliably. But this is not what the comment says - it suggests that the "disruption" (the close) doesn't always work, but according to the HTTPClientError you checked below, it seems it does work every time.

nuivall replied:
I suspect following reasons:
- perhaps scylla is sending bytes before we are able to return error in python, for instance it lands in some OS queue or python library queue first, in such case error doesn't always affect the flow 
- it's not a single connection in the driver but a pool (they say default is 10 I think but it's difficult to track in their code as it's very "dynamic", e.g. converts some strings to class names and instantiates it, stuff like that).

nuivall replied:
second point is maybe less relevant here,  as you noticed it client side error always happens

nuivall replied:
although maybe if the connection is closed then reusing it significantly increases crash chance

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2236/

nyh commented:
> > Unfortunately, even with the new version, I can't seem to get the test to ever fail without your code patch...
> 
> That's odd, it reproduces 100% on my machine. I have scylla master [ff386e7](https://github.com/scylladb/scylladb/commit/ff386e7a445d17ed2199a26e4ac75905b071a91b) and boto3:
> 
> ```
> ❯ pip show boto3
> Name: boto3
> Version: 1.20.26
> ```

Something really weird happens to me today. If I compile Scylla and run the new test (test/alternator/run test_batch.py::test_batch_write_item_large_broken_connection), it fails with an assertion failure exactly like you wanted to show. But if I then run it again, many times, it always passes! I can't think of any way to explain this. This is SO WEIRD - I repeat this strange observation three times - "git reset --hard; ninja" to recompile Scylla, then run the test, it fails, then run the tests 10 more times, they all pass. I can't explain it, but ran out of energy to try to explain what's going on.

So I'll merge this version as-is. The code patch is definitely correct. And should be backported everywhere we can.
